feature_engine/variable_manipulation.py

```diff
@@ -127,6 +135,116 @@ def _find_or_check_categorical_variables(
     return variables
 
 
+def _convert_variable_to_datetime(X: pd.Series, **kwargs) -> pd.Series:
+    """
+    Take a series and attempt to convert its dtype to datetime unless
+    its type is numeric. In that case, return the series as is.
+    If the series type is object or category, it is processed as follows:
+        1 if it could be cast as number, return the series as is
+        2 otherwise, attempt to convert it to datetime; if pd.to_datetime
+          fails the conversion, return the series as is
+
+    Arguments
+    ---------
+    X: pd.Series that may or may not contain values convertible to datetime
+    kwargs: see pd.to_datetime()
+
+    **Notes**
+    if the series type is category, it is the dtype of the categories
+    that is converted, not the series itself This means that the primary type
+    is still going to be category whereas the pd.Series.dtype.categories
+    property is going to have its type converted.
+
+    Missing values in str/cat objects are going to be replaced by 'NaT'
+    in the converted datetime columns
+    **
+    """
+
+    if hasattr(X, "dt"):
+        return X
+
+    if isinstance(X.dtype, (pd.CategoricalDtype, object)):
+        try:
+            X.astype("float64")
+            return X
+        except ValueError:
+            return X.apply(pd.to_datetime, errors="ignore", **kwargs)
+
+    elif np.issubdtype(X.dtype, np.number):
+        return X
+
+    return X.apply(pd.to_datetime, errors="ignore", **kwargs)
+
+
+def _convert_variables_to_datetime(
+    X: pd.DataFrame, variables: Variables = None, **kwargs
+) -> pd.DataFrame:
+    """
+    Takes a dataframe and returns it with object/category features
+    that represent a datetime converted into datetime variables.
+    It does not manipulate numerical features or object/category features
+    that could be casted into numerical type (see _convert_variable_to_datetime)
+    If variables == None, process all the columns in the dataframe.
+    """
+
+    variables = X.columns if not variables else variables
+    X_dt = X.copy()
+    for var in X_dt[variables].select_dtypes(exclude="number"):
+        X_dt[var] = _convert_variable_to_datetime(X[var], **kwargs)
+
+    return X_dt
+
+
+def _find_or_check_datetime_variables(
+    X: pd.DataFrame, variables: Variables = None
+) -> List[Union[str, int]]:
+    """
+    Checks that variables provided by the user are of type datetime,
+    and transform date/times given in str/obj format into datetimes.
+    If None, finds all the datetime type variables in the DataFrame.
+
+    Parameters
+    ----------
+    X : pandas DataFrame
+    variables : variable or list of variables. Defaults to None.
+
+    Raises
+    ------
+    ValueError
+        If there are no datetime variables in df or df is empty
+    TypeError
+        If any of the user provided variables are not datetime or
+        [failed to convert str/obj to datetime]?
+
```

Comment:
I would structure this function more or less like this:

For variables indicated by the user in a list:

if they are not type timestamp or object, return an error
for those that are object, attempt to convert to datetime, if fails, return an error

if user does not indicate which variables, then we should find them:
step 1: find those variables in the df that are of timestamp dtype, 
step 2: find those that are object that can be convertible to dt

Reply:
This part is a little more iffy, it is definitely the one that took me the longest to figure out. Unfortunately I think it still needs a bit more discussing

The way it currently works is more or less as you outlined above. However:
- Timestamp is the type of a __single__ datetime observation, not a column/series. After calling pd.to_datetime(), the series.dtype is _usually_ np.datetime64
- The reason why I say _usually_ is because when the data is timezone aware, the dtype is something slightly different and pandas has a _duck type_ for it called pd.DatetimeTZDtype. Also, there might be cases where datetimes object get cast to something else that I haven't come across.
- Therefore, instead of checking the _type_ of the datetime variables, I decided to check whether they have the __dt__ attribute, which happens to be the attribute used by the ExtractDateTransformer to extract the features afterwards. I believe all datetime specific dtypes have in fact the attribute dt; in addition, if some obscure datetime format ever shows up that doesn't have the dt attribute, it would be intended behaviour to reject it as the transformer wouldn't be able to work on such variables anyway.

To sum up on _find_or_check_datetime_variables:

if the user indicates the variables:

1. pass those variables to _convert_variables_to_datetime, which will attempt to convert non-numerical variables to datetime. The variables for which a conversion was not possible will be returned unchanged by passing errors='ignore' to pd.to_datetime() calls

2. if after the conversion some variables still do not have the dt attribute (which, I reiterate, is our argument to determine that they are in fact datetime variables that the downstream transformers can work with), throw a TypeError

if the user does not indicate the variables:

1. same as above, but pass the whole dataframe
2. if no variables have the dt attribute after the conversion, throw an error

In both cases, if no errors are thrown, return the variables that have the dt attribute

Reply:
Sorry about the textwall but I couldn't really be more concise than this I don't think. I found this specific piece of the code quite cumbersome to go through. What do you think?

Reply:
did you consider pandas [s_datetime64_any_dtype](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.api.types.is_datetime64_any_dtype.html)? I found it in this [stackoverflow thread](https://stackoverflow.com/questions/43214204/how-do-i-tell-if-a-column-in-a-pandas-dataframe-is-of-type-datetime-how-do-i-te).

Not sure it would work but I would try something like:

`from pandas.api.types import is_datetime64_any_dtype as is_datetime`

If user does not indicate list:
```
vars_df = [column for column in df.columns if is_datetime(df[column])]]
vars_ob = [column for column in df.columns if is_datetimepd.to_dattime((df[column]))]
variables = vars_df + vars_ob
```

If user indicates list:
```
var_ = c for c in df.olumns if not is_datetime(df[column] or not is_datetimepd.to_dattime((df[column])
if len(var_)>0 ==> return error
```

Reply:
Thanks for raising this.

To start of, I would just consider object. I think this would capture the great majority of cases. And make our lives easier :)